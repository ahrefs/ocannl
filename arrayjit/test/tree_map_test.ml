open Base
open Stdio
open Utils

(* Demo of Tree_map with tree-preserving sexp output *)
let () =
  let open Tree_map in

  (* Create a tree by inserting values *)
  let tree =
    empty
    |> add ~compare:Int.compare ~key:5 ~data:"five"
    |> add ~compare:Int.compare ~key:3 ~data:"three"
    |> add ~compare:Int.compare ~key:7 ~data:"seven"
    |> add ~compare:Int.compare ~key:2 ~data:"two"
    |> add ~compare:Int.compare ~key:4 ~data:"four"
    |> add ~compare:Int.compare ~key:6 ~data:"six"
    |> add ~compare:Int.compare ~key:8 ~data:"eight"
  in

  (* Print the tree structure *)
  printf "Tree structure:\n";
  printf "%s\n\n" (Sexp.to_string_hum (sexp_of_t Int.sexp_of_t String.sexp_of_t tree));

  (* Test lookups *)
  printf "Finding key 4: %s\n" (Option.value ~default:"not found" (find ~compare:Int.compare ~key:4 tree));
  printf "Finding key 10: %s\n" (Option.value ~default:"not found" (find ~compare:Int.compare ~key:10 tree));

  (* Print as association list *)
  printf "\nAs alist (in-order): ";
  printf "%s\n" (Sexp.to_string_hum ([%sexp_of: (int * string) list] (to_alist tree)))
