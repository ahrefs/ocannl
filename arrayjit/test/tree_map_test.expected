Retrieving commandline, environment, or config file variable ocannl_log_level
Found 0, in the config file
Tree structure:
(Node (key 5) (value five)
 (left
  (Node (key 3) (value three)
   (left (Node (key 2) (value two) (left Empty) (right Empty)))
   (right (Node (key 4) (value four) (left Empty) (right Empty)))))
 (right
  (Node (key 7) (value seven)
   (left (Node (key 6) (value six) (left Empty) (right Empty)))
   (right (Node (key 8) (value eight) (left Empty) (right Empty))))))

Finding key 4: four
Finding key 10: not found

As alist (in-order): ((2 two) (3 three) (4 four) (5 five) (6 six) (7 seven) (8 eight))
